### PR TITLE
Support Complex number

### LINKED ISSUE #127: Support complex type

### PR DESCRIPTION
**Pull request by okuta**
This PR based on chainer/chainer#1543 and #214.
Fix #127.

### CODE REVIEW

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -1533,7 +1570,8 @@ cdef class ndarray:
             raise TypeError('{} array cannot be set to {} array'.format(
                 arr.dtype, self.dtype))
         if self.shape != arr.shape:
-            raise ValueError('Shape mismatch')
+            raise ValueError('Shape mismatch. Old shape: {}, new shape: {}'
+                             ''.format(self.shape, arr.shape))
```
Comment (unnonouno) [inline cupy/core/core.pyx]: This style is a little bit curious

File: cupy/core/core.pyx [hunk 2]
```diff
@@ -3529,11 +3590,71 @@ add = create_arithmetic(
...
+    .. seealso:: :data:`numpy.imag`
+
+    ''')
+
+
+_imag_setter = create_ufunc(
+    'cupy_imag_setter',
+    ('f->F', 'd->D'),
+    'out0.imag(in0)',
+    doc='''Sets the imag part of the elements of the array.
```
Comment (unnonouno) [inline cupy/core/core.pyx]: imaginary

File: cupy/core/core.pyx [hunk 3]
```diff
@@ -3529,11 +3590,71 @@ add = create_arithmetic(
...
+    'out0.real(in0)',
+    doc='''Sets the real part of the elements of the array.
+    ''')
+
+
+imag = create_ufunc(
+    'cupy_imag',
+    ('F->f', 'D->d'),
+    'out0 = in0.imag()',
+    doc='''Return the imaginary part of the elements of the array.
```
Comment (unnonouno) [inline cupy/core/core.pyx]: Returns

File: cupy/core/core.pyx [hunk 4]
```diff
@@ -3529,11 +3590,71 @@ add = create_arithmetic(
...
+    .. seealso:: :data:`numpy.angle`
+
+    ''')
+
+
+real = create_ufunc(
+    'cupy_real',
+    ('F->f', 'D->d'),
+    'out0 = in0.real()',
+    doc='''Return the real part of the elements of the array.
```
Comment (unnonouno) [inline cupy/core/core.pyx]: Returns

File: cupy/core/core.pyx [hunk 5]
```diff
@@ -3529,11 +3590,71 @@ add = create_arithmetic(
...
+    .. seealso:: :data:`numpy.conj`
+
+    ''')
+
+
+angle = create_ufunc(
+    'cupy_angle',
+    ('F->f', 'D->d'),
+    'out0 = arg(in0)',
+    doc='''Return the angle of the complex argument.
```
Comment (unnonouno) [inline cupy/core/core.pyx]: Returns

File: cupy/core/core.pyx [hunk 6]
```diff
@@ -3529,11 +3590,71 @@ add = create_arithmetic(
     ''')
 
 
+conj = create_ufunc(
+    'cupy_conj',
+    ('F->F', 'D->D'),
+    'out0 = conj(in0)',
+    doc='''Return the complex conjugate, element-wise.
```
Comment (unnonouno) [inline cupy/core/core.pyx]: Returns

File: cupy/core/core.pyx [hunk 7]
```diff
@@ -3712,7 +3844,11 @@ cdef _mean = create_reduction_func(
     ('?->d', 'B->d', 'h->d', 'H->d', 'i->d', 'I->d', 'l->d', 'L->d',
      'q->d', 'Q->d',
      ('e->e', (None, None, None, 'float')),
-     'f->f', 'd->d'),
+     'f->f', 'd->d',
+     ('F->F', ('in0', 'a + b',
+               'out0 = a / float(_in_ind.size() / _out_ind.size())', None)),
```
Comment (unnonouno) [inline cupy/core/core.pyx]: `(float)`?

File: cupy/testing/helper.py [hunk 8]
```diff
@@ -500,20 +502,33 @@ def test_func(self, *args, **kw):
 _dtypes = _float_dtypes + _int_bool_dtypes
 
 
-def _make_all_dtypes(no_float16, no_bool):
-    if no_float16:
-        if no_bool:
-            return _regular_float_dtypes + _int_dtypes
+def _make_all_dtypes(no_float16, no_bool, no_complex):
```
Comment (unnonouno) [inline cupy/testing/helper.py]: Can you make simpler?

```
dtypes = []
...
if not no_complex:
    dtypes += _complex_dtypes
```

**Comment by unnonouno** [comment]
How do you think about test for imaginary part of complex numbers? Arryas `shaped_arange` generates have imaginary part?

**Comment by okuta** [comment]
I fixed `shaped_arange`.

File: cupy/testing/helper.py [hunk 9]
```diff
@@ -879,11 +879,13 @@ def shaped_arange(shape, xp=cupy, dtype=numpy.float32):
...
     """
+    dtype = numpy.dtype(dtype).type
     a = numpy.arange(1, internal.prod(shape) + 1, 1)
-    if numpy.dtype(dtype).type == numpy.bool_:
-        return xp.array((a % 2 == 0).reshape(shape))
-    else:
-        return xp.array(a.astype(dtype).reshape(shape))
+    if dtype == numpy.bool_:
+        a = a % 2 == 0
+    elif dtype == numpy.complex64 or dtype == numpy.complex128:
```
Comment (unnonouno) [inline cupy/testing/helper.py]: How about using `.kind == 'c'`?

File: cupy/testing/helper.py [hunk 10]
```diff
@@ -870,11 +879,13 @@ def shaped_arange(shape, xp=cupy, dtype=numpy.float32):
...
+    dtype = numpy.dtype(dtype)
     a = numpy.arange(1, internal.prod(shape) + 1, 1)
-    if numpy.dtype(dtype).type == numpy.bool_:
-        return xp.array((a % 2 == 0).reshape(shape))
-    else:
-        return xp.array(a.astype(dtype).reshape(shape))
+    if dtype.type == numpy.bool_:
+        a = a % 2 == 0
+    elif dtype.kind == 'c':
+        a = a + ([0, 1j, -1j] * (len(a) // 3 + 1))[0:len(a)]
```
Comment (unnonouno) [inline cupy/testing/helper.py]: How about using `a + a * 1j`?

File: cupy/testing/helper.py [hunk 11]
```diff
@@ -870,11 +879,13 @@ def shaped_arange(shape, xp=cupy, dtype=numpy.float32):
          ``True`` (resp. ``False``).
 
     """
+    dtype = numpy.dtype(dtype)
     a = numpy.arange(1, internal.prod(shape) + 1, 1)
-    if numpy.dtype(dtype).type == numpy.bool_:
-        return xp.array((a % 2 == 0).reshape(shape))
-    else:
-        return xp.array(a.astype(dtype).reshape(shape))
+    if dtype.type == numpy.bool_:
```
Comment (unnonouno) [inline cupy/testing/helper.py]: `dtype == '?'`?

File: cupy/testing/helper.py [hunk 12]
```diff
@@ -893,12 +904,14 @@ def shaped_reverse_arange(shape, xp=cupy, dtype=numpy.float32):
...
     size = internal.prod(shape)
     a = numpy.arange(size, 0, -1)
-    if numpy.dtype(dtype).type == numpy.bool_:
-        return xp.array((a % 2 == 0).reshape(shape))
-    else:
-        return xp.array(a.astype(dtype).reshape(shape))
+    if dtype.type == numpy.bool_:
+        a = a % 2 == 0
+    elif dtype.kind == 'c':
+        a = a + ([0, 1j, -1j] * (len(a) // 3 + 1))[0:len(a)]
```
Comment (unnonouno) [inline cupy/testing/helper.py]: ditto

**Comment by okuta** [comment]
Please merge #367 before this.

**Comment by okuta** [comment]
I resolved conflicts.

File: tests/cupy_tests/random_tests/test_generator.py [hunk 13]
```diff
@@ -31,6 +31,8 @@ def __exit__(self, *_):
 
 
 @testing.fix_random()
+@unittest.skipUnless(
+    os.sys.platform != 'win32', 'dtype problem on Windows')
```
Comment (okuta) [inline tests/cupy_tests/random_tests/test_generator.py]: I will create another PR.

**Comment by unnonouno** [comment]
LGTM

**Pull request merged**